src/Patcher.cs

```diff
@@ -0,0 +1,362 @@
+                isCacheUpdateNeeded = true;
+            }
+        }
+
+        int alreadyHadMMHOOKCount = 0;
+        int totalValidPluginsCount = currentPlugins.Count;
+
+        for(int i = 0; i < currentPlugins.Count; i++)
+        {
+            var plugin = currentPlugins[i];
+            if(plugin.IsDuplicate){
+                ExtendedLogging($"[{nameof(Begin)}] Skipping HookGen Check For Duplicate or Old Version of {plugin.GUID} ({plugin.PluginVersion})");
+                totalValidPluginsCount--;
+                continue;
+            }
+
+            if(plugin.BadImageFormat){
+                ExtendedLogging($"[{nameof(Begin)}] Skipping HookGen Check For {plugin.GUID} Because {nameof(BadImageFormatException)}");
+                totalValidPluginsCount--;
+                continue;
+            }
+
+            if(plugin.AlreadyHasMMHOOK){
+                alreadyHadMMHOOKCount++;
+                // ExtendedLogging($"[{nameof(Begin)}] Assembly Already Has MMHOOK: " + plugin.GUID);
+                continue;
+            }
+
+            if(BoundConfig.GenerateForAllPlugins.Value || mmhookOverrides.Contains(plugin.GUID) || currentPlugins.SelectMany(plugins => plugins.References).Contains(plugin.GUID))
+            {
+                var hookGenTask = new Task(() => StartHookGen(plugin));
+                hookGenTask.Start();
+                hookGenTasks.Add(hookGenTask);
+            }
+        }
+
+        if(BoundConfig.GenerateForAllPlugins.Value && BoundConfig.DisableGenerateForAllPlugins.Value)
+        {
+            Logger.LogInfo($"[{nameof(Begin)}] DisableGenerateForAllPlugins is enabled, disabling GenerateForAllPlugins.");
+            BoundConfig.GenerateForAllPlugins.Value = false;
+        }
+        
+        ExtendedLogging($"[{nameof(Begin)}] Already Has MMHOOK: {alreadyHadMMHOOKCount} Out of {totalValidPluginsCount} Assemblies");
+        ExtendedLogging($"[{nameof(Begin)}] Waiting for {hookGenTasks.Count} HookGen Task{(hookGenTasks.Count == 1 ? "" : "s")} to finish.");
+
+        hookGenTasks.ForEach(x => x.Wait());
+        
+        if(isCacheUpdateNeeded || cachedPlugins == null)
+            WriteCache(currentPlugins);
+    }
+
+    private static void StartHookGen(CachedAssemblyInfo plugin)
+    {
+        if(HookGenPatch.RunHookGen(plugin)){
+            lock(plugin){
+                plugin.AlreadyHasMMHOOK = true;
+            }
+            isCacheUpdateNeeded = true;
+        }
+    }
+
+    private static IEnumerable<CachedAssemblyInfo> GetPlugins(List<CachedAssemblyInfo>? cachedAssemblies)
+    {
+        List<string> paths = new();
+        Directory.GetFiles(Paths.ManagedPath, "*.dll", SearchOption.AllDirectories).ToList().ForEach(paths.Add);
+        Directory.GetFiles(Paths.BepInExAssemblyDirectory, "*.dll", SearchOption.AllDirectories).ToList().ForEach(paths.Add);
+        Directory.GetFiles(Paths.PatcherPluginPath, "*.dll", SearchOption.AllDirectories).ToList().ForEach(paths.Add);
```

Comment:
Patchers should be unloaded during runtime, so never accept mmhook files for them

Reply:
Hmm, I see. Yeah maybe it doesn't really make *too* much sense to have it, but then again there could be some rare case where it could be useful, but yeah I don't know if that case will ever come

Reply:
Anyways about this, I've thought about it, and I think this functionality should stay because the whole point of this software is to generate MMHOOK files for the user. And we can't know what the user (modder) is doing and as I said before, it's entirely possible they want to patch a patcher from their plugin, as we also can't know what kind of patchers there might be.

I don't like placing restrictions on what kind of assemblies can be MMHOOK'd if there is no technical reason for it. Well, I don't actually know if there is a technical reason for this, so do correct me if there is.